src/components/DebounceInput.tsx

```diff
@@ -0,0 +1,119 @@
+import React from 'react';
+import {
+  setDisplayName,
+  compose,
+  defaultProps,
+  pure,
+  withProps,
+  lifecycle,
+  withState,
+  withHandlers,
+} from 'recompose';
+import {
+  WrappedFieldProps,
+} from 'redux-form';
+import debounce from 'lodash.debounce';
+import * as r from 'ramda';
+
+type OnChange = (evt: React.ChangeEvent<HTMLInputElement>) => any;
+
+interface Indexed {
+  [key: string]: any;
+}
+
+export interface OuterProps extends WrappedFieldProps, Indexed {
+  wait?: number;
+  ownerComponent: React.ComponentType<WrappedFieldProps & Indexed>,
+}
+
+export interface DefaultProps extends OuterProps {
+  wait: number;
+}
+
+export interface StateProps extends DefaultProps {
+  debounceFieldValue: string;
+  setDebounceFieldValue: (v: string) => any,
+  debouncing: boolean,
+  setDebouncing: (v: boolean) => any,
+}
+
+export interface HandlerProps extends StateProps {
+  onChange: OnChange,
+}
+
+export interface InnerProps extends WrappedFieldProps, Indexed {
+  ownerComponent: React.ComponentType<WrappedFieldProps & Indexed>,
+}
+
+const PureDebounceInput = ({
+  ownerComponent: Component,
+  wait,
+  debounceFieldValue,
+  setDebounceFieldValue,
+  onChange,
+  ...props
+}: InnerProps) => (
+  <Component {...props}/>
+);
+
+const enhance = compose<InnerProps, OuterProps>(
+  setDisplayName('DebounceInput'),
+  pure,
+  defaultProps({ wait: 250 }),
+  withState('debounceFieldValue', 'setDebounceFieldValue', ''),
```

Comment:
If you have multiple `withState` hocs think of using: https://github.com/acdlite/recompose/blob/master/docs/API.md#withstatehandlers

Reply:
done